### PR TITLE
HAL-09: add custom errors to Well.sol

### PR DESCRIPTION
**Pull request by silochad**
@pynchmeister:
1. Implement these errors in Well.sol (see https://blog.soliditylang.org/2021/04/21/custom-errors/)
2. Update all tests to use the custom errors (see https://book.getfoundry.sh/cheatcodes/expect-revert#examples)
3. Add Natspec docs to each error

### CODE REVIEW

File: src/Well.sol [hunk 1]
```diff
@@ -212,7 +212,7 @@ contract Well is ERC20PermitUpgradeable, IWell, ReentrancyGuardUpgradeable, Clon
         // slippage from imprecision goes to the Well or to the User.
         amountOut = reserveJBefore - reserves[j];
 
-        require(amountOut >= minAmountOut, "Well: slippage");
+        if(!(amountOut >= minAmountOut)) revert SlippageOut(amountOut, minAmountOut);
```
Comment (silochad) [inline src/Well.sol:215]:
```suggestion
        if(amountOut < minAmountOut) revert SlippageOut(amountOut, minAmountOut);
```

File: src/Well.sol [hunk 2]
```diff
@@ -252,7 +252,7 @@ contract Well is ERC20PermitUpgradeable, IWell, ReentrancyGuardUpgradeable, Clon
 
         // slippage check needs to move to the actual amount out
         // need to take delta across _executeSwap
-        require(amountIn <= maxAmountIn, "Well: slippage");
+        if(!(amountIn <= maxAmountIn)) revert SlippageIn(amountIn, maxAmountIn);
```
Comment (silochad) [inline src/Well.sol:255]:
```suggestion
        if(amountIn > maxAmountIn) revert SlippageIn(amountIn, maxAmountIn);
```

File: src/Well.sol [hunk 3]
```diff
@@ -327,7 +327,7 @@ contract Well is ERC20PermitUpgradeable, IWell, ReentrancyGuardUpgradeable, Clon
             }
         }
         lpAmountOut = _calcLpTokenSupply(wellFunction(), reserves) - totalSupply();
-        require(lpAmountOut >= minLpAmountOut, "Well: slippage");
+        if(!(lpAmountOut >= minLpAmountOut)) revert SlippageOut(lpAmountOut, minLpAmountOut);
```
Comment (silochad) [inline src/Well.sol:330]:
```suggestion
        if(lpAmountOut < minLpAmountOut) revert SlippageOut(lpAmountOut, minLpAmountOut);
```

File: src/Well.sol [hunk 4]
```diff
@@ -390,7 +390,7 @@ contract Well is ERC20PermitUpgradeable, IWell, ReentrancyGuardUpgradeable, Clon
         uint j = _getJ(_tokens, tokenOut);
 
         tokenAmountOut = _getRemoveLiquidityOneTokenOut(lpAmountIn, j, reserves);
-        require(tokenAmountOut >= minTokenAmountOut, "Well: slippage");
+        if(!(tokenAmountOut >= minTokenAmountOut)) revert SlippageOut(tokenAmountOut, minTokenAmountOut);
```
Comment (silochad) [inline src/Well.sol:393]:
```suggestion
        if(tokenAmountOut < minTokenAmountOut) revert SlippageOut(tokenAmountOut, minTokenAmountOut);
```

File: src/Well.sol [hunk 5]
```diff
@@ -441,7 +441,7 @@ contract Well is ERC20PermitUpgradeable, IWell, ReentrancyGuardUpgradeable, Clon
             reserves[i] = reserves[i] - tokenAmountsOut[i];
         }
         lpAmountIn = totalSupply() - _calcLpTokenSupply(wellFunction(), reserves);
-        require(lpAmountIn <= maxLpAmountIn, "Well: slippage");
+        if(!(lpAmountIn <= maxLpAmountIn)) revert SlippageIn(lpAmountIn, maxLpAmountIn);
```
Comment (silochad) [inline src/Well.sol:444]:
```suggestion
        if(lpAmountIn > maxLpAmountIn) revert SlippageIn(lpAmountIn, maxLpAmountIn);
```

File: src/Well.sol [hunk 6]
```diff
@@ -577,7 +577,7 @@ contract Well is ERC20PermitUpgradeable, IWell, ReentrancyGuardUpgradeable, Clon
      */
     function _setReserves(IERC20[] memory _tokens, uint[] memory reserves) internal {
         for (uint i; i < reserves.length; ++i) {
-            require(reserves[i] <= _tokens[i].balanceOf(address(this)), "Well: Invalid reserve");
+            if(!(reserves[i] <= _tokens[i].balanceOf(address(this)))) revert InvalidReserves();
```
Comment (silochad) [inline src/Well.sol:580]:
```suggestion
            if(reserves[i] > _tokens[i].balanceOf(address(this))) revert InvalidReserves();
```

File: src/interfaces/IWell.sol [hunk 7]
```diff
@@ -16,6 +16,17 @@ struct Call {
...
+    /// @dev Returns only a uint amountOut and a uint minAmountOut.
+    error SlippageOut(uint amountOut, uint minAmountOut);
+    /// @notice Returns the slippage of swapping `amountIn` of `fromToken` for `toToken`.
+    /// @dev Returns only a uint amounntIn and a uint maxAmountIn.
+    error SlippageIn(uint amountIn, uint maxAmountIn);
+    /// @notice Returns the amount of leaves the tree has.
+    error InvalidTokens();
+    /// @notice Returns the amount of leaves the tree has.
+    error InvalidReserves();
+
```
Comment (silochad) [inline src/interfaces/IWell.sol:29]: Please use the same block comment format for consistency.

File: src/interfaces/IWell.sol [hunk 8]
```diff
@@ -16,6 +16,17 @@ struct Call {
  * @title IWell is the interface for the Well contract.
  */
 interface IWell {
+    /// @notice Returns the slippage of swapping `amountOut` of `fromToken` for `toToken`.
```
Comment (silochad) [inline src/interfaces/IWell.sol:19]: Let's shoot for a more descriptive error message here.

This does not return slippage; this error is thrown when the `amountOut` doesn't pass the `minAmountOut` threshold.

```suggestion
    /// @notice Thrown when an operation would deliver fewer tokens than `minAmountOut`.
```

File: src/interfaces/IWell.sol [hunk 9]
```diff
@@ -16,6 +16,17 @@ struct Call {
...
+    /// @dev Returns only a uint amountOut and a uint minAmountOut.
+    error SlippageOut(uint amountOut, uint minAmountOut);
```
Comment (silochad) [inline src/interfaces/IWell.sol:21]: Please add a line break between each error for readability

File: src/interfaces/IWell.sol [hunk 10]
```diff
@@ -16,6 +16,17 @@ struct Call {
...
+    /// @notice Returns the slippage of swapping `amountIn` of `fromToken` for `toToken`.
+    /// @dev Returns only a uint amounntIn and a uint maxAmountIn.
+    error SlippageIn(uint amountIn, uint maxAmountIn);
+    /// @notice Returns the amount of leaves the tree has.
```
Comment (silochad) [inline src/interfaces/IWell.sol:25]: Is this ChatGPT? Let's update this to describe what actually happens. Refer to the `_getIJ` function.

**Comment by silochad** [comment]
We need to update all tests of Well.sol which revert to use the custom error message.

Running `yarn test` now produces a lot of errors, like this:

```
[FAIL. Reason: Error != expected error: 0x672215de != Well: Invalid tokens] test_getIJ_revertIfBothMissing() (gas: 14730)
```

File: src/interfaces/IWell.sol (hunk 7 above)
Comment (pynchmeister) [reply src/interfaces/IWell.sol:29]: Fair enough

File: src/interfaces/IWell.sol (hunk 10 above)
Comment (pynchmeister) [reply src/interfaces/IWell.sol:25]: No just human error, I'll get right on this.

**Review by k-xo — APPROVED** [review]

**Pull request merged**